File: api/app/services/shopping_actions.py
import json
import os
import re

from colorama import Fore, Style
from google import genai
from sqlalchemy import delete, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import joinedload

from app.db.models import ShoppingListItem, ShoppingProduct
from app.prompts.shopping_prompt import (
    get_confirm_product_match_prompt,
    get_delete_product_prompt,
    get_new_product_prompt,
    get_product_match_prompt,
)
from app.utils.conversation_state import (
    clear_conversation_status,
    get_conversation_status,
    set_conversation_status,
)

client = genai.Client(api_key=os.getenv("GEMINI_API_KEY"))
model = os.getenv("BASIC_AI_MODEL") or "gemini-2.0-flash"


async def handle_add_to_shopping_list(
    message: str, discord_id: str, db: AsyncSession
) -> str:
    product_list_raw = await db.execute(select(ShoppingProduct))
    products = product_list_raw.scalars().all()
    product_list_text = "\n".join([f"- id:{p.id} {p.name}" for p in products])

    prompt = get_product_match_prompt(product_list_text, message)
    response = await client.aio.models.generate_content(model=model, contents=prompt)
    print(Fore.GREEN + "AI response:" + str(response.text) + Style.RESET_ALL)

    clean = re.sub(r"```(?:json)?\n?", "", response.text or "").strip()
    data = json.loads(clean)
    print(Fore.YELLOW + "AI data:" + str(data) + Style.RESET_ALL)

    if data.get("status") == "confirm":
        set_conversation_status(
            discord_id, {"state": "awaiting_confirm", "product_id": data["product_id"]}
        )
        return (
            f"Nie jestem pewna, czy '{message}' to wskazany produkt. "
            f"Czy chcesz dodać ten produkt do listy zakupów? (tak/nie)"
        )

    if data.get("status") == "not_found":
        set_conversation_status(
            discord_id, {"state": "awaiting_new_product", "attempts": 0}
        )
        return (
            "Nie znalazłam tego produktu. Jeśli chcesz dodać nowy produkt do bazy, "
            "podaj jego nazwę."
        )

    if data.get("status") == "found":
        product_id = data.get("product_id")
        if product_id is None:
            return "Nie udało się dopasować produktu. Spróbuj jeszcze raz."

        existing = await db.execute(
            select(ShoppingListItem).where(ShoppingListItem.product_id == product_id)
        )
        if existing.scalar_one_or_none():
            return "Ten produkt jest już na liście zakupów!"

        item = ShoppingListItem(product_id=product_id)
        db.add(item)
        await db.commit()
        product_result = await db.execute(
            select(ShoppingProduct).where(ShoppingProduct.id == product_id)
        )
        product = product_result.scalar_one_or_none()
        if not product:
            return "Dodano produkt do listy zakupów."
        return f"Dodano '{product.name}' do listy zakupów!"

    return "Nie udało się dopasować produktu. Spróbuj jeszcze raz."


async def handle_create_new_product(
    text: str, discord_id: str, db: AsyncSession
) -> str:
    print(Fore.RED + "message" + str(text) + Style.RESET_ALL)

    prompt = get_new_product_prompt(text)
    response = await client.aio.models.generate_content(model=model, contents=prompt)
    print(Fore.GREEN + "AI response:" + str(response.text) + Style.RESET_ALL)

    clean = re.sub(r"```(?:json)?\n?", "", response.text or "").strip()
    data = json.loads(clean)
    new_product_name = (data.get("name") or "").strip()

    if data.get("status") == "cancelled":
        clear_conversation_status(discord_id)
        return "Ok, anulowano."

    if data.get("status") == "invalid":
        state = get_conversation_status(discord_id) or {}
        attempts = int(state.get("attempts", 0)) + 1

        if attempts >= 3:
            clear_conversation_status(discord_id)
            return "Nowy produkt nie został dodany do listy."

        set_conversation_status(
            discord_id, {"state": "awaiting_new_product", "attempts": attempts}
        )
        return (
            "To nie jest produkt, który można kupić w sklepie. "
            "Podaj samą nazwę produktu."
        )

    if data.get("status") == "valid":
        if not new_product_name:
            return "Nie udało się odczytać nazwy produktu. Spróbuj jeszcze raz."

        new_product = ShoppingProduct(name=new_product_name)
        db.add(new_product)
        await db.commit()
        await db.refresh(new_product)

        item = ShoppingListItem(product_id=new_product.id)
        db.add(item)
        await db.commit()

        clear_conversation_status(discord_id)
        return f"Dodałem '{new_product_name}' do bazy i do listy zakupów!"

    return "Nie udało się rozpoznać odpowiedzi. Spróbuj jeszcze raz."


async def handle_awaiting_confirm(
    message: str, discord_id: str, product_id: int, db: AsyncSession
) -> str:
    prompt = get_confirm_product_match_prompt(message)
    response = await client.aio.models.generate_content(model=model, contents=prompt)
    print(Fore.GREEN + "AI response:" + str(response.text) + Style.RESET_ALL)

    clean = re.sub(r"```(?:json)?\n?", "", response.text or "").strip()
    data = json.loads(clean)
    decision = data.get("decision")

    if decision == "confirm":
        existing = await db.execute(
            select(ShoppingListItem).where(ShoppingListItem.product_id == product_id)
        )
        if existing.scalar_one_or_none():
            return "Ten produkt jest już na liście zakupów!"

        item = ShoppingListItem(product_id=product_id)
        db.add(item)
        await db.commit()
        clear_conversation_status(discord_id)

        product_result = await db.execute(
            select(ShoppingProduct).where(ShoppingProduct.id == product_id)
        )
        product = product_result.scalar_one_or_none()
        if not product:
            return "Dodano produkt do listy zakupów."
        return f"Dodano '{product.name}' do listy zakupów!"

    if decision == "new_product":
        clear_conversation_status(discord_id)
        new_message = (data.get("product_text") or message).strip()
        return await handle_add_to_shopping_list(new_message, discord_id, db)

    clear_conversation_status(discord_id)
    return "Ok, nie dodałam tego produktu."


async def handle_clear_shopping_list(db: AsyncSession) -> str:
    await db.execute(delete(ShoppingListItem))
    await db.commit()
    return "Lista zakupów została wyczyszczona!"


async def handle_get_shopping_list(db: AsyncSession) -> str:
    result = await db.execute(
        select(ShoppingListItem).options(joinedload(ShoppingListItem.product))
    )
    items = result.scalars().all()
    if not items:
        return "Lista zakupów jest pusta!"
    product_names = [str(item.product.name) for item in items if item.product]
    return "Lista zakupów:\n" + "\n".join([f"- {name}" for name in product_names])


async def handle_delete_from_shopping_list(text: str, db: AsyncSession) -> str:
    product_list_raw = await db.execute(select(ShoppingProduct))
    products = product_list_raw.scalars().all()
    product_list_text = "\n".join([f"- id:{p.id} {p.name}" for p in products])

    prompt = get_delete_product_prompt(product_list_text, text)
    response = await client.aio.models.generate_content(model=model, contents=prompt)
    print(Fore.GREEN + "AI response:" + str(response.text) + Style.RESET_ALL)

    clean = re.sub(r"```(?:json)?\n?", "", response.text or "").strip()
    data = json.loads(clean)

    product_id = data.get("product_id")
    if product_id is None:
        return "Taki produkt nie istnieje"

    existing = await db.execute(
        select(ShoppingListItem).where(ShoppingListItem.product_id == product_id)
    )
    product = existing.scalar_one_or_none()

    if product:
        product_result = await db.execute(
            select(ShoppingProduct).where(ShoppingProduct.id == product_id)
        )
        product_entity = product_result.scalar_one_or_none()

        await db.execute(delete(ShoppingListItem).where(ShoppingListItem.product_id == product_id))
        await db.commit()

        if product_entity:
            return f"Produkt '{product_entity.name}' został usunięty z listy zakupów!"
        return "Produkt został usunięty z listy zakupów!"

    return "Nie ma tego produktu na liście zakupów!"

File: api/src/services/shoppingActions.ts
import { GoogleGenAI } from "@google/genai";
import { prisma } from "../db/prisma";
import {
  getProductMatchPrompt,
  getConfirmProductMatchPrompt,
  getNewProductPrompt,
  getDeleteProductPrompt,
} from "../prompts/shoppingPrompt";
import {
  setConversationStatus,
  getConversationStatus,
  clearConversationStatus,
} from "../utils/conversationState";
import { MESSAGES } from "../messages";

const ai = new GoogleGenAI({ apiKey: process.env["GEMINI_API_KEY"] });
const model = process.env["BASIC_AI_MODEL"] ?? "gemini-2.0-flash";

function parseJson(text: string) {
  return JSON.parse(text.replace(/```(?:json)?\n?/g, "").trim());
}

export async function handleAddToShoppingList(message: string, discordId: string): Promise<string> {
  const products = await prisma.shoppingProduct.findMany();
  const productListText = products.map((p) => `- id:${p.id} ${p.name}`).join("\n");

  const response = await ai.models.generateContent({ model, contents: getProductMatchPrompt(productListText, message) });
  console.log("[add_to_shopping] AI:", response.text);

  const data = parseJson(response.text ?? "{}");

  if (data.status === "confirm") {
    setConversationStatus(discordId, { state: "awaiting_confirm", product_id: data.product_id });
    return MESSAGES.CONFIRM_PRODUCT_MATCH(message);
  }

  if (data.status === "not_found") {
    setConversationStatus(discordId, { state: "awaiting_new_product", attempts: 0 });
    return MESSAGES.PRODUCT_NOT_FOUND_ASK_NEW;
  }

  if (data.status === "found") {
    const productId: number = data.product_id;
    if (!productId) return MESSAGES.PRODUCT_MATCH_FAILED;

    const existing = await prisma.shoppingListItem.findFirst({ where: { productId } });
    if (existing) return MESSAGES.PRODUCT_ALREADY_ON_LIST;

    await prisma.shoppingListItem.create({ data: { productId } });
    const product = await prisma.shoppingProduct.findUnique({ where: { id: productId } });
    return product ? MESSAGES.PRODUCT_ADDED(product.name) : MESSAGES.PRODUCT_ADDED_FALLBACK;
  }

  return MESSAGES.PRODUCT_MATCH_FAILED;
}

export async function handleCreateNewProduct(text: string, discordId: string): Promise<string> {
  console.log("[create_new_product] message:", text);

  const response = await ai.models.generateContent({ model, contents: getNewProductPrompt(text) });
  console.log("[create_new_product] AI:", response.text);

  const data = parseJson(response.text ?? "{}");

  if (data.status === "cancelled") {
    clearConversationStatus(discordId);
    return MESSAGES.CANCELLED;
  }

  if (data.status === "invalid") {
    const state = getConversationStatus(discordId) ?? {} as Record<string, unknown>;
    const attempts = (Number(state["attempts"]) || 0) + 1;

    if (attempts >= 3) {
      clearConversationStatus(discordId);
      return MESSAGES.NEW_PRODUCT_NOT_ADDED;
    }

    setConversationStatus(discordId, { state: "awaiting_new_product", attempts });
    return MESSAGES.PRODUCT_NOT_VALID;
  }

  if (data.status === "valid") {
    const name = (data.name as string)?.trim();
    if (!name) return MESSAGES.PRODUCT_NAME_READ_FAILED;

    const newProduct = await prisma.shoppingProduct.create({ data: { name } });
    await prisma.shoppingListItem.create({ data: { productId: newProduct.id } });
    clearConversationStatus(discordId);
    return MESSAGES.NEW_PRODUCT_ADDED(name);
  }

  return MESSAGES.PRODUCT_RESPONSE_UNRECOGNIZED;
}

export async function handleAwaitingConfirm(message: string, discordId: string, productId: number): Promise<string> {
  const response = await ai.models.generateContent({ model, contents: getConfirmProductMatchPrompt(message) });
  console.log("[awaiting_confirm] AI:", response.text);

  const data = parseJson(response.text ?? "{}");

  if (data.decision === "confirm") {
    const existing = await prisma.shoppingListItem.findFirst({ where: { productId } });
    if (existing) return MESSAGES.PRODUCT_ALREADY_ON_LIST;

    await prisma.shoppingListItem.create({ data: { productId } });
    clearConversationStatus(discordId);

    const product = await prisma.shoppingProduct.findUnique({ where: { id: productId } });
    return product ? MESSAGES.PRODUCT_ADDED(product.name) : MESSAGES.PRODUCT_ADDED_FALLBACK;
  }

  if (data.decision === "new_product") {
    clearConversationStatus(discordId);
    const newMessage = ((data.product_text as string) || message).trim();
    return handleAddToShoppingList(newMessage, discordId);
  }

  clearConversationStatus(discordId);
  return MESSAGES.DECLINED;
}

export async function handleClearShoppingList(): Promise<string> {
  await prisma.shoppingListItem.deleteMany();
  return MESSAGES.SHOPPING_LIST_CLEARED;
}

export async function handleGetShoppingList(): Promise<string> {
  const items = await prisma.shoppingListItem.findMany({ include: { product: true } });
  if (!items.length) return MESSAGES.SHOPPING_LIST_EMPTY;
  const names = items.map((i) => i.product.name);
  return MESSAGES.SHOPPING_LIST_HEADER + "\n" + names.map((n) => `- ${n}`).join("\n");
}

export async function handleDeleteFromShoppingList(text: string): Promise<string> {
  const products = await prisma.shoppingProduct.findMany();
  const productListText = products.map((p) => `- id:${p.id} ${p.name}`).join("\n");

  const response = await ai.models.generateContent({ model, contents: getDeleteProductPrompt(productListText, text) });
  console.log("[delete_from_shopping] AI:", response.text);

  const data = parseJson(response.text ?? "{}");
  const productId: number | null = data.product_id ?? null;
  if (!productId) return MESSAGES.PRODUCT_NOT_IN_DB;

  const item = await prisma.shoppingListItem.findFirst({ where: { productId } });
  if (!item) return MESSAGES.PRODUCT_NOT_ON_LIST;

  const product = await prisma.shoppingProduct.findUnique({ where: { id: productId } });
  await prisma.shoppingListItem.deleteMany({ where: { productId } });

  return product ? MESSAGES.PRODUCT_DELETED(product.name) : MESSAGES.PRODUCT_DELETED_FALLBACK;
}
